jbmc/src/java_bytecode/convert_java_nondet.cpp

```diff
@@ -39,71 +39,83 @@ static goto_programt::targett insert_nondet_init_code(
-  if(!lhs.type().has_subtype())
-  {
-    return next_instr;
-  }
+    const typet &type = side_effect.type();
+    // If the lhs type doesn't have a subtype then I guess it's primitive and
+    // we want to bail out now
+    if(!(type.id() == ID_pointer && type.has_subtype()))
+    {
+      continue;
+    }
 
-  // Although, if the type is a ptr-to-void then we also want to bail
-  if(lhs.type().subtype().id()==ID_empty ||
-     lhs.type().subtype().id()==ID_code)
-  {
-    return next_instr;
-  }
+    // Although, if the type is a ptr-to-void then we also want to bail
+    if(type.subtype().id() == ID_empty || type.subtype().id() == ID_code)
+    {
+      continue;
+    }
 
-  // Check whether the nondet object may be null
-  if(!to_side_effect_expr_nondet(side_effect).get_nullable())
-    object_factory_parameters.max_nonnull_tree_depth = 1;
-  // Get the symbol to nondet-init
-  const auto source_loc=target->source_location;
-
-  // Erase the nondet assignment
-  target->make_skip();
-
-  // Generate nondet init code
-  code_blockt init_code;
-  gen_nondet_init(
-    lhs,
-    init_code,
-    symbol_table,
-    source_loc,
-    true,
-    allocation_typet::DYNAMIC,
-    object_factory_parameters,
-    update_in_placet::NO_UPDATE_IN_PLACE);
-
-  // Convert this code into goto instructions
-  goto_programt new_instructions;
-  goto_convert(
-    init_code, symbol_table, new_instructions, message_handler, mode);
-
-  // Insert the new instructions into the instruction list
-  goto_program.destructive_insert(next_instr, new_instructions);
-  goto_program.update();
+    // Check whether the nondet object may be null
+    if(!to_side_effect_expr_nondet(side_effect).get_nullable())
+      object_factory_parameters.max_nonnull_tree_depth = 1;
+    // Get the symbol to nondet-init
+    const auto source_loc = target->source_location;
+
+    // Create aux symbol for nondet object
+    symbolt &aux_symbol = get_fresh_aux_symbol(
+      type,
+      id2string(goto_programt::get_function_id(goto_program)),
+      "nondet_tmp",
+      source_loc,
+      ID_java,
+      symbol_table);
+    aux_symbol.is_static_lifetime = false;
+
+    const symbol_exprt aux_symbol_expr = aux_symbol.symbol_expr();
+    code_declt decl(aux_symbol_expr);
+    decl.add_source_location() = source_loc;
+
+    // Generate nondet init code
+    code_blockt init_code;
+    init_code.add(decl);
+
+    gen_nondet_init(
+      aux_symbol_expr,
+      init_code,
+      symbol_table,
+      source_loc,
+      true,
+      allocation_typet::DYNAMIC,
+      object_factory_parameters,
+      update_in_placet::NO_UPDATE_IN_PLACE);
+
+    // Convert this code into goto instructions
+    goto_programt new_instructions;
+    goto_convert(
+      init_code, symbol_table, new_instructions, message_handler, mode);
+
+    // Insert the new instructions into the instruction list before target
+    goto_program.destructive_insert(target, new_instructions);
+    goto_program.update();
```

Comment:
The `update()` procedure is quite expensive, so best done en masse after all replacements. The one caveat: instructions won't necessarily have a reasonable `location_number` before `update()` is run, but this pass does not appear to use `location_number`.